run docstrings tests with mocha

aspeddro opened the pull request:
Main changes:

- Generates a single file with all the examples from the docstrings, `generated_mocha_test.res`
- Runtime tests using `mocha`
- Much faster
- Run on macos and Linux

aspeddro committed:
run tests with mocha

aspeddro committed:
trigger ci

aspeddro committed:
remove ci branch

aspeddro committed:
test on all platform

aspeddro committed:
fix windows cli path

aspeddro committed:
fix test

aspeddro committed:
skip windows

aspeddro committed:
order output

aspeddro committed:
fix order

aspeddro committed:
cleanup

aspeddro commented on scripts/test.js:
```diff
@@ -136,14 +136,28 @@ async function runTests() {
-      //   },
-      // );
+
+      cp.execSync(`node ${path.join("tests", "docstrings_examples", "DocTest.res.mjs")}`, {
+        cwd: path.join(__dirname, ".."),
+        stdio: [0, 1, 2],
+      })
```
It will create the file `generated_mocha_test.res`

cknitt replied:
Could you add this as a comment in the code?

aspeddro replied:
Done

aspeddro commented on scripts/test.js:
```diff
@@ -136,14 +136,28 @@ async function runTests() {
+        stdio: [0, 1, 2],
+      })
+
+      cp.execSync(`${rescript_exe} build`, {
+        cwd: path.join(__dirname, "..", "tests/docstrings_examples"),
+        stdio: [0, 1, 2],
+      });
```
Run build again to check compile errors

cknitt replied:
Could you add this as a comment in the code?

aspeddro replied:
Done

cknitt commented on tests/docstrings_examples/DocTest.res:
```diff
@@ -9,24 +9,19 @@ type example = {
-| None => []
-}
+// Ignore some tests not supported by node v18
+let ignoreRuntimeTests = [
```
Bonus points: detect node version and run tests accordingly. As these tests actually work with Node 22.

aspeddro replied:
Done.

cknitt commented on tests/docstrings_examples/DocTest.res:
```diff
@@ -158,142 +153,45 @@ let extractExamples = async () => {
+let main = async () => {
+  let examples = await extractExamples()
+  examples->Array.sort((a, b) =>
+    Obj.magic(a.id) > Obj.magic(b.id) ? Ordering.fromInt(1) : Ordering.fromInt(-1)
```
Why do we need `Obj.magic` here?

I think this can just be
```rescript
  examples->Array.sort((a, b) => String.compare(a.id, b.id))
```
and should be moved into the `extractExamples` function.

aspeddro replied:
Thanks

cknitt reviewed (commented):
Great work! This is awesome. Huge speedup (> 10x on my machine).

Some feedback:

1. `generated_mocha.test.res` is quite huge. Maybe we shouldn't check it in?
2. If possible, could we change the generation so that we get one `describe` call per module (e.g., `String`) with multiple `test` calls inside (one per function, e.g., `charAt`, `charCodeAt`, ...).
3. Is the inner `module Test` actually needed? In what cases?

aspeddro committed:
adjusts

aspeddro commented:
> Great work! This is awesome. Huge speedup (> 10x on my machine).
> 
> Some feedback:
> 
> 1. `generated_mocha.test.res` is quite huge. Maybe we shouldn't check it in?
> 2. If possible, could we change the generation so that we get one `describe` call per module (e.g., `String`) with multiple `test` calls inside (one per function, e.g., `charAt`, `charCodeAt`, ...).
> 3. Is the inner `module Test` actually needed? In what cases?

1. Added `generated_mocha_test.res` and `generated_mocha_test.res.mjs` to gitignore.
2. Done
3. Yes, bc some examples contains type definitions. Added comment on code

aspeddro committed:
Remove `generated_mocha_test.res` if file exists

aspeddro committed:
format

cknitt commented:
> 2. Done

Doesn't really seem to be done yet? I will see one `describe` per `test`.

But never mind, we can always improve things in separate PRs.

cknitt commented:
> Doesn't really seem to be done yet? I still see one `describe` per `test`.

Sorry, please disregard, seems I somehow had a stale generated file locally. 🤦
